### PR TITLE
Public apiserver should check binding from service broker before poli…

### PR DESCRIPTION
**Pull request by qibobo**
…cy relevant operations.

[#151906019]

### CODE REVIEW

File: api/app.js [hunk 1]
```diff
@@ -68,7 +68,21 @@ module.exports = function(settings, callback) {
...
-
+  var checkBinding = function(req, res, next){
+    serviceBrokerUtil.checkBinding(req.params,function(error,result){
+      if(error){
+        res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({});
+      }else{
+        if(result.statusCode == HttpStatus.OK){
+          next();
+        }else if(result.statusCode == HttpStatus.NOT_FOUND){
+          res.status(HttpStatus.NOT_FOUND).send({"message": "binding_info_not_found"});
```
Comment (cdlliuy) [inline api/app.js]: Suggest to return 403 forbidden with msg "The request cannot be performed without an existing service binding"

File: servicebroker/lib/server.js [hunk 2]
```diff
@@ -59,32 +84,53 @@ module.exports = function(settings, catalog, callback) {
...
+        });
+    }
 
-    var server;
+    var internalServer;
     if(settings.tls){
-        server = https.createServer(options, app).listen(port, function() {
-            var port = server.address().port;
-            logger.info('Service broker app is running in secure mode', { port: port });
+        internalServer = https.createServer(options, internalApp).listen(port, function() {
```
Comment (cdlliuy) [inline servicebroker/lib/server.js]: Please help to take care of the  indentation

File: api/test/unit/routes/policyHelper.test.js [hunk 3]
```diff
@@ -37,6 +37,9 @@ describe('Policy Route helper ', function() {
 		});
   	})
 	beforeEach(function() {
+		nock(serviceBrokerURI)
```
Comment (cdlliuy) [inline api/test/unit/routes/policyHelper.test.js]: Please help to take care of the  indentation

File: api/app.js [hunk 4]
```diff
@@ -68,7 +68,21 @@ module.exports = function(settings, callback) {
...
-
+  var checkBinding = function(req, res, next){
+    serviceBrokerUtil.checkBinding(req.params,function(error,result){
+      if(error){
+        res.status(HttpStatus.INTERNAL_SERVER_ERROR).send({});
+      }else{
+        if(result.statusCode == HttpStatus.OK){
+          next();
+        }else if(result.statusCode == HttpStatus.NOT_FOUND){
+          res.status(HttpStatus.FORBIDDEN).send({"message": "The request cannot be performed without an existing service binding"});
```
Comment (boyang9527) [inline api/app.js]: change the message to "the application is not bound to  Auto-Scaling service "

File: api/lib/utils/serviceBrokerUtils.js [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    var appId = parameters.app_id;
+    logger.info("Checking binding", { "appId": appId });
+    var options = getOptions(appId);
+    request(options, function(error, response, body) {
+      if (error) {
+        logger.error("Error occurred during check binding ", { "appId": appId, "error": error });
+        error.statusCode = HttpStatus.INTERNAL_SERVER_ERROR;
+        callback(error, null);
+      } else if (response.statusCode === HttpStatus.OK) {
+        logger.info("Binding info exists", { "appId": appId });
```
Comment (boyang9527) [inline api/lib/utils/serviceBrokerUtils.js]: change "Binding info exists" to "Service Binding exists"

File: api/lib/utils/serviceBrokerUtils.js [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        callback(null, { "statusCode": HttpStatus.OK });
+      } else if (response.statusCode === HttpStatus.NOT_FOUND) {
+        if (body && body.message && body.message === "binding_info_not_found") {
```
Comment (boyang9527) [inline api/lib/utils/serviceBrokerUtils.js]: why check body?

File: servicebroker/lib/internalRoutes/binding.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+
+  app.get('/v1/apps/:app_id/service_bindings', function(req, res) {
+    var appId = req.params.app_id;
+    if (appId == null || appId.trim() === "") {
+      logger.error("appId is required");
+      res.status(400).json({});
+      return;
+    }
+    models.binding.count({ where: { appId: appId } }).then(function(count) {
```
Comment (boyang9527) [inline servicebroker/lib/internalRoutes/binding.js]: return the binding information not matter the count is 0 or not.

File: servicebroker/lib/internalRoutes/binding.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+      if (count > 0) {
+        res.status(200).json({"message": "binding_info_exists"});
+        return;
+      } else {
+        res.status(404).json({ "message": "binding_info_not_found" });
```
Comment (boyang9527) [inline servicebroker/lib/internalRoutes/binding.js]: don't return 404 not found if there is no binding.

**Comment by qibobo** [comment]
@boyang9527  updated.

**Comment by boyang9527** [comment]
@cdlliuy pls review  Yang's change.

**Comment by cdlliuy** [comment]
LGTM

**Pull request merged**